librz/type/parser/c_cpp_parser.c

```diff
@@ -0,0 +1,288 @@
+// SPDX-FileCopyrightText: 2021 Anton Kochkov <anton.kochkov@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <stdio.h>
+#include <rz_types.h>
+#include <rz_list.h>
+#include <rz_util/rz_file.h>
+#include <rz_type.h>
+#include <tree_sitter/api.h>
+
+#include <types_parser.h>
+
+// Declare the `tree_sitter_c` function, which is
+// implemented by the `tree-sitter-c` library.
+TSLanguage *tree_sitter_c();
+
+// Declare the `tree_sitter_cpp` function, which is
+// implemented by the `tree-sitter-cpp` library.
+//TSLanguage *tree_sitter_cpp();
+
+CParserState *c_parser_state_new(HtPP *base_types, HtPP *callable_types) {
+	CParserState *state = RZ_NEW0(CParserState);
+	if (!base_types) {
+		state->types = ht_pp_new0();
+	} else {
+		state->types = base_types;
+	}
+	if (!callable_types) {
+		state->callables = ht_pp_new0();
+	} else {
+		state->callables = callable_types;
+	}
+	// Forward definitions require to have a special hashtable
+	state->forward = ht_pp_new0();
+	state->errors = rz_strbuf_new("");
+	state->warnings = rz_strbuf_new("");
+	state->debug = rz_strbuf_new("");
+	return state;
+}
+
+void c_parser_state_free(CParserState *state) {
+	ht_pp_free(state->forward);
+	ht_pp_free(state->types);
+	ht_pp_free(state->callables);
+	rz_strbuf_free(state->debug);
+	rz_strbuf_free(state->warnings);
+	rz_strbuf_free(state->errors);
+	free(state);
+	return;
+}
+
+void c_parser_state_free_keep_ht(CParserState *state) {
+	rz_strbuf_free(state->debug);
+	rz_strbuf_free(state->warnings);
+	rz_strbuf_free(state->errors);
+	free(state);
+	return;
+}
+
+struct rz_type_parser_t {
+	CParserState *state;
+};
+
+RZ_API RZ_OWN RzTypeParser *rz_type_parser_new() {
+	RzTypeParser *parser = RZ_NEW0(RzTypeParser);
+	if (!parser) {
+		return NULL;
+	}
+	parser->state = c_parser_state_new(NULL, NULL);
+	return parser;
+}
+
+RZ_API RZ_OWN RzTypeParser *rz_type_parser_init(HtPP *types, HtPP *callables) {
+	RzTypeParser *parser = RZ_NEW0(RzTypeParser);
+	if (!parser) {
+		return NULL;
+	}
+	parser->state = c_parser_state_new(types, callables);
+	return parser;
+}
+
+RZ_API void rz_type_parser_free(RZ_NONNULL RzTypeParser *parser) {
+	// We do not destroy HT by default since it might be used after
+	c_parser_state_free_keep_ht(parser->state);
+	free(parser);
+}
+
+RZ_API void rz_type_parser_free_purge(RZ_NONNULL RzTypeParser *parser) {
+	c_parser_state_free(parser->state);
+	free(parser);
+}
+
+static int type_parse_string(CParserState *state, const char *code, char **error_msg) {
+	// Create a parser.
+	TSParser *parser = ts_parser_new();
+	// Set the parser's language (C in this case)
+	ts_parser_set_language(parser, tree_sitter_c());
+
+	TSTree *tree = ts_parser_parse_string(parser, NULL, code, strlen(code));
+
+	// Get the root node of the syntax tree.
+	TSNode root_node = ts_tree_root_node(tree);
+	int root_node_child_count = ts_node_named_child_count(root_node);
```

Comment:
When do you expect this to happen?

Reply:
If it will be just `(translation_unit)`, without any nodes.